### PR TITLE
aws: Configure Karpenter resources in the cluster spec

### LINKED ISSUE #15046: karpenter-controller CPU request is way too high and not spread across nodes
In the addon, karpenter-controller reserves [half a CPU](https://github.com/kubernetes/kops/blob/3bafdc414288b7525bbdf67e93d6ccfa5bde091d/upup/models/cloudup/resources/addons/karpenter.sh/k8s-1.19.yaml.template#L883). Looking at my own stats, it barely needs more than 2%.  

image

Additionally the controller deployment is using the wrong `topologySpreadConstraint` key (`webhook` instead of `controller`): https://github.com/kubernetes/kops/blob/3bafdc414288b7525bbdf67e93d6ccfa5bde091d/upup/models/cloudup/resources/addons/karpenter.sh/k8s-1.19.yaml.template#L948-L960

### PR DESCRIPTION
**Pull request by danports**
Allow configuration of the CPU request, memory request, and memory limit for Karpenter pods in the cluster spec.

This PR also changes the default request for CPU from 500m to 25m and for memory from 1Gi to 200Mi (the memory limit is unchanged at 1Gi). Is this too harsh? Would it be better to err on the side of higher requests and risk wasting resources on smaller clusters or err on the side of lower requests and risk starving pods on larger clusters?

Closes #15046.

### CODE REVIEW

File: upup/models/cloudup/resources/addons/karpenter.sh/k8s-1.19.yaml.template [hunk 1]
```diff
@@ -1526,10 +1526,10 @@ spec:
               port: http
           resources:
             limits:
-              memory: 1Gi
+              memory: {{ or .Karpenter.MemoryLimit "1Gi" }}
             requests:
-              cpu: 500m
-              memory: 1Gi
+              cpu: {{ or .Karpenter.CPURequest "25m" }}
+              memory: {{ or .Karpenter.MemoryRequest "200Mi" }}
```
Comment (hakman) [inline upup/models/cloudup/resources/addons/karpenter.sh/k8s-1.19.yaml.template:1532]: Let's be more conservative for now.
```suggestion
              cpu: {{ or .Karpenter.CPURequest "200m" }}
              memory: {{ or .Karpenter.MemoryRequest "500Mi" }}
```

**Review by hakman — CHANGES_REQUESTED** [review]

**Comment by danports** [comment]
Test failure is suspicious because it has "karpenter" in the name, but reviewing the logs, it seems like the cluster deployed just fine but the test flaked out because validation didn't complete in time - hopefully it passes on the second try:
```
I0915 10:11:45.761542   37599 validate_cluster.go:220] (will retry): cluster passed validation 7 consecutive times
Error: validation failed: wait time exceeded during validation
```

**Comment by danports** [comment]
I'm puzzled. The Karpenter tests seem to be failing fairly consistently (though the `pull-kops-e2e-aws-upgrade-k126-ko127-to-k127-kolatest-karpenter` test didn't fail the first time around), but I don't see anything obvious wrong in the test logs - the Karpenter pods themselves seem to be running just fine:
```
  Sep 15 20:33:14.145: INFO: karpenter-6965dbc8fd-qmt8s started at 2023-09-15 20:02:20 +0000 UTC (0+1 container statuses recorded)
  Sep 15 20:33:14.145: INFO: 	Container controller ready: true, restart count 0
```
Is there something I'm missing here? The log is ~31K lines so it's quite possible. I could try reverting the change I made to the default Karpenter resource requests to see if they were set that way for the sake of the tests.

**Comment by hakman** [comment]
Let's focus on getting pull-kops-e2e-aws-karpenter to pass.
/test pull-kops-e2e-aws-karpenter

**Comment by hakman** [comment]
I need to investigate when things started to fail.
https://testgrid.k8s.io/kops-misc#kops-aws-karpenter

**Comment by danports** [comment]
Alright. I tried reverting the change I made to the default Karpenter resource requests and that didn't help, so it doesn't seem likely the failures are related to this PR. Based on that test grid link, it seems like `pull-kops-e2e-aws-karpenter` has been mostly failing since early August, was passing for a couple weeks in late July, and was failing before that.

**Comment by danports** [comment]
The `pull-kops-e2e-aws-karpenter` test started failing in August after the upgrade from Cilium v1.12.4 to v1.13.5 was merged: https://github.com/kubernetes/kops/commit/ba7b4e716acca2ced8bf9a1b72ebcd0e5b769319

That may be related since the cluster the test deploys runs on Cilium - though I imagine if that were the only cause other tests would also be failing?

**Comment by hakman** [comment]
Please rebase and we can merge.
/lgtm

**Review by hakman — APPROVED** [review]

**Review by hakman — APPROVED** [review]

**Pull request merged**